File: akka-mapreduce-scala/src/main/scala/cn/v5cn/akka/mapreduce_scala/actors/MasterActor.scala
package cn.v5cn.akka.mapreduce_scala.actors

import akka.actor.Actor
import cn.v5cn.akka.mapreduce_scala.MapData
import cn.v5cn.akka.mapreduce_scala.ReduceData
import cn.v5cn.akka.mapreduce_scala.Result
import akka.actor.Props
import akka.routing.RoundRobinPool
import akka.routing.RoundRobinPool

class MasterActor extends Actor {
  
  val mapActor = context.actorOf(RoundRobinPool(nrOfInstances=5).props(Props[MapActor]),name="map")
  val reduceActor = context.actorOf(RoundRobinPool(nrOfInstances=5).props(Props[ReduceActor]),name="reduce")
  val aggregateActor = context.actorOf(Props[AggregateActor],name="aggregate")
  
	def receive = {
	  case line:String =>  mapActor ! line
	  case mapData:MapData => reduceActor ! mapData
	  case reduceData:ReduceData => aggregateActor ! reduceData
	  case Result => aggregateActor forward Result
	}
}

File: akka-mapreduce-java/src/main/java/cn/v5cn/akka/mapreduce_java/actors/MasterActor.java
package cn.v5cn.akka.mapreduce_java.actors;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.routing.RoundRobinPool;
import cn.v5cn.akka.mapreduce_java.messages.MapData;
import cn.v5cn.akka.mapreduce_java.messages.ReduceData;
import cn.v5cn.akka.mapreduce_java.messages.Result;

public class MasterActor extends UntypedActor {
	
	ActorRef mapActor = getContext().actorOf(new RoundRobinPool(5).props(Props.create(MapActor.class)),"map");//getContext().actorOf(Props.create(MapActor.class).withRouter(new RoundRobinRouter(5)),"map");
	ActorRef reduceActor = getContext().actorOf(new RoundRobinPool(5).props(Props.create(ReduceActor.class)),"reduce");//.withRouter(new RoundRobinRouter(5)),"reduce");
	ActorRef aggregateActor = getContext().actorOf(Props.create(AggregateActor.class),"aggregate");
	
	@Override
	public void onReceive(Object message) throws Exception {
		if(message instanceof String){
			mapActor.tell(message, getSelf());
		}else if(message instanceof MapData){
			reduceActor.tell(message, getSelf());
		}else if(message instanceof ReduceData){
			aggregateActor.tell(message, ActorRef.noSender());
		}else if(message instanceof Result){
			aggregateActor.forward(message, getContext());
		}else{
			unhandled(message);
		}
	}

}
